### PR TITLE
Pattern: InnerSource License

### PR DESCRIPTION
**Pull request by spier**
InnerSource License pattern. Implements #138.

This has already gone through 3 iterations between Cornelius and myself (Sebastian).
We tried to use the Pattern template to the best of our knowledge.

We would like to get this merged into master as quickly as possible, so that Cornelius can get feedback from a broader audience, and hopefully also hear from companies that have used similar approaches. 

If it speeds up the merge to master, we can also post this Pattern as "Pattern Drafts (proven, not yet fully reviewed)".

### CODE REVIEW

**Comment by spier** [comment]
@cornelius this is the result of our work in the gDoc, converted to markdown. Pending review from others in the InnerSourceCommons Patterns community.

**Comment by spier** [comment]
@lenucksi how would I best go about finding a reviewer for this PR?

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]:
```suggestion
An organization faces legal uncertainty about InnerSource collaboration among the multiple legal entities its composed of. An **InnerSource License** makes explicit the rights and obligations of those a legal entities providing and using the component.```

Comment (maxcapraro) [reply innersource-license.md]: This is technically two proposed changes in one - both open for discussion obviously:

**First)** I'd propose to use `organization` composed of multiple `legal entities`:

- *Organization* - an organization in a broad and less narrowly defined sense (e.g. Lufthansa)
- *Legal entity* - an entity that has own legal rights and obligations (e.g. Lufthansa Systems GmbH, Lufthansa Industry Solutions TS GmbH, ...)

My worry is that both `company` and `group` are highly ambiguous terms. Readers might have different ideas of what a group or company is. 

Another I had in mind was to borrow terminology from tax accounting and talk about `associated enterprises` and plainly other, non-associated enterprises. However, this doesn't help us in distinguishing the fine-grained parts (legal entities) from the whole (organization).

**Second)** I adapted (shortened) the patlet to be more focused on what - I think - are the key insights of the pattern.

Comment (maxcapraro) [reply innersource-license.md]: (I'll not point out the first suggestion again to not clutter the PR ... I am happy to provide commit suggestions once we read a consensus though)

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]: I'd like to dig a little deeper on the root causes for the `... they need ...` part and make them more explicit. What are the reasons?

- Is it a fear of illegal profit shifting between the legal entities?
- Is it a general accounting issue / obligation? (i.e. you need to make explicit rights and obligations to properly record liabilities and new assets that InnerSource-style sharing brings to the different legal entities)
- Is it something else?

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md:14]: Could be `When two or more ...` (Can't make commit suggestion because I do not want to mix the two threads)

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md:11]: The section describes the problem of legal uncertainty really well. I feel there is a second (related) problem that the license solves: 

In addition to the legal uncertainty, there is the a psychological uncertainty: "There is this piece of code. Can I really just use this for my work?" The license resolves this as well. This is similar to what licenses in open source (and particularly in the context free software) achieved: They spread the gospel of collaboration.

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]:
```suggestion
This uncertainty creates an effective barrier to InnerSource collaboration among the different legal entities of the organization.
```

I feel the fact that the patterns in this repo are "InnerSource Patterns" establishes the context that we want to bring down silos, avoid duplication, forster reuse and collaboration. I'd suggest we not list them again to avoid duplication.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md:11]: (On the other side: This could distract readers from the key problem at hand - the legal requirements and the barriers they can introduce)

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]: This is an excellent description of why this problem is a problem 👍

Comment (maxcapraro) [reply innersource-license.md]: Some sugestion for word smithing (feel free to commit, steal parts, or ignore)
```suggestion
At the time of sharing the source code, it can not be reliably predicted what the value of sharing will be. Hence if the activity of sharing creates significant effort (i.e. negotiating terms for the usage), the companies are less likely to do it: They cannot predict if the effort will pays off.
```

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]:
```suggestion
- These legal entities are entities with own legal rights, obligations
```

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: (Digging could happen as part of the forces, too)

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]:
```suggestion
```

Doesn't that apply to InnerSource in general? Is it a contextual factor only important for this pattern?

Comment (maxcapraro) [reply innersource-license.md]: (Mental note to ourselves: Maybe we need to define a general context (some kind of an `AbstractSuperContext`) for all patterns)

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]: My first reaction was: "Well, that's a general InnerSource force, so maybe we should drop it." But the more I think about it, I think it was right to list it here: Missing routine with sharing / collaborating, might increase the need for a license.

Maybe we should add a sentence though, to explain how this force ties into the pattern?

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]: `Internal regulation` are one thing. Another thing are plain legal requirements. Did you interviews yield any insights on this (potential) additional force?

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md:52]: I konw of two other organizations using an MIT-based license. I'll contact them and ask if they would want to provide a paragraph about it here or at least allow us to write about it.

This shouldn't stop this PR from being merged though - we can always add it later.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md:11]: Yippie! That goes into the direction of what you observed at DB (further down): 

> One positive effect that is already showing is that it simplifies the conversation, especially if some of the involved parties don’t know the InnerSource concept that well yet. Licenses are a well-known concept, therefore having an InnerSource License is a great discussion starter.

File: innersource-license.md
Comment (maxcapraro) [inline innersource-license.md]: Can we beef that up somehow? 

- What are thing people should be aware of when applying this pattern?
- What were positive side effects observed?
- What were negative side effects observed? (I always feel that is the greatest part of the resulting context)

**Review by maxcapraro — CHANGES_REQUESTED** [review]
Wow. I am really excited about this pattern. Great great great work, @spier!

I added some thoughts and comments. At times, I get a bit pedantic about writing texts (habbit from my day job) - so please stop me where I went overboard. 

One additional thing that would be great: Do you think you could win Schlomo or Cornelius to also perform a review? :)

**Comment by spier** [comment]
Thanks. I will take a look at your feedback. 

Cornelius has written this thing with me ;) So I don’t think he needs to review it again.

**Comment by maxcapraro** [comment]
> Cornelius has written this thing with me ;) So I don’t think he needs to review it again.

True that :) Congrats on this very good initial version to you as well, Cornelius!

File: innersource-license.md
Comment (spier) [reply innersource-license.md:52]: Input from other companies with similar models will certainly be interesting. Just as you said, let's move that out of the context of this PR though.

If you want, you can even file a new issue that says "Harden InnerSource License pattern with experience from x and y".

Will resolve this without action.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: We did not go deeper into this. The **internal regulations** aspect seemed to be the more pressing one.

@cornelius what do you say, do **external regulations** also impact this pattern? Should we add a bullet about **external regulations** to the **Forces**?

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I agree that there are likely some general points that always apply when an InnerSource approach is considered by a company.

For the time being I would like to keep this in here though, as we don't have such a general **InnerSource Context** (`AbstractSuperContext`) description yet. My assumption is that the readers need to be able to "read the pattern on it's own". 

But I can see that if we ever want to turn all proven patterns into a book, there would be an intro section that describes the **InnerSource Context**.

Do you want to create an issue to describe that **InnerSource Context** (`AbstractSuperContext`)? @gruetter could be interested in that. At the summit the topic of "how large does a company have to be for InnerSource to make sense" came up in a breakout that we were in.

@maxcapraro please close this conversation if you are fine with leaving this section as is.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Just like our other thread about the **InnerSource Context**, I would like to keep this in here to allow the readers to understand the pattern without having to read anything else.

@maxcapraro please close this conversation if you are fine with leaving this section as is.

File: innersource-license.md
Comment (spier) [reply innersource-license.md:14]: Great suggestion. Have added this in.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Cornelius and I also went back and forth between calling it *enterprise* or *group*. And we landed at *group* eventually.

For the time being I suggest to keep it as *group* so that we don't have to reword everything. Also I don't know which terms are easier to understand, or which ones are more correct in legal speak.

Maybe when other companies that have similar challenges add their experience to this pattern, then we can test which term they prefer?

Comment (spier) [reply innersource-license.md]: I really like the edits you did to the Patlet!
I will take parts of that but keep the group/company terms as they are for the time being.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I think what you are looking for is mostly described in **Known Instances**. There we tried to describe the observations that are specific to DB Systel. Also as they have barely rolled this InnerSource License concept out, the experience is not that hardened yet. 

Maybe we can review this **Resulting Context** again in a couple of months or together with other companies that have applied this pattern too?

I would not change this for now.

@maxcapraro please close this conversation if you are fine with leaving this section as is.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Do you mean the whole section you marked or only **Company culture**?

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I was only able to work in part of that, without changing the original meaning of the Patlet too much.

Will keep this conversation open also for reference for Cornelius.

File: innersource-license.md
Comment (spier) [reply innersource-license.md:11]: I am unclear if you are suggesting any change here?

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: The Patlet section contains

> they are concerned about the implications in terms of legal liabilities or cross-company accounting

@maxcapraro are you suggesting that we elaborate on that part in the Problem section more?

**Comment by spier** [comment]
@maxcapraro thanks for the detailed review.

Unfortunately there were only a few "easy merges" in your suggestions so I often had to ask clarifying questions.

To get this PR merged to master I would suggest these next steps:

- please check my feedback above and resolve the threads where you are happy to leave things as they are (to bring down the number of open threads to a reasonable level)
- we mention Cornelius explicitly on open threads where we want his input on. After a deadline (say 2 weeks) we take what we have at that point and merge the "best possible iteration" of this into master
- if you want to harden this pattern through a review by companies that have applied similar patterns, I would suggest opening a new issue for that. That would become easier as well once this PR is in master
- if you want to pull out the **InnerSource Context** (`AbstractSuperContext`) from the patterns, I would suggest opening a separate issue for that too

What do you think?

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I worked in something similar. Resolving.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Great feedback. I worked in something similar.

File: innersource-license.md
Comment (cornelius) [reply innersource-license.md]: Regarding the motivation I think there are two high-level root causes:
* There are legal contracts in place which cover sharing source code but they are not standardized (so they create additional effort in negotiating and understanding for every project) or they don't allow sharing source code in an open enough sense as an Inner Source approach would suggest.
* There are no contracts in place but source code is shared informally. That might create uncertainty in cases where clarity about ownership and rights and obligations is needed.

Accounting, profit sharing, taxes are all things which could be covered in an Inner Source license, but they also could be left to other legal constructions.

Ownership of code, relicensing, liability, warranty are topics you probably would typically cover in a license.

In summary I would say the goal is to create clarity about the terms under which source code is shared.

File: innersource-license.md
Comment (cornelius) [reply innersource-license.md]: I think the specific condition for an Inner Source License is that the organization has a complexity so that there are legal contracts necessary within the organization. That probably corresponds with size and numbers of developers in most cases.

File: innersource-license.md
Comment (cornelius) [reply innersource-license.md]: I would say that external regulations do have an impact, but not a very specific one. Legal documents have to match plain legal requirements, but that's kind of a given.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: > For the time being I suggest to keep it as group so that we don't have to reword everything.

I know writing is a lot of work and time is a scarce resource. Let's still try to find the best terminology possible. We / I can support with the rewording once we settled on terms.

> Maybe when other companies that have similar challenges add their experience to this pattern, then we can test which term they prefer?

In the organizations I worked with, we ended up using exclusively `legal entity` to talk about the - well - legal entities.

Comment (maxcapraro) [reply innersource-license.md]: Summary of arguments, both gathered from this discussion and from my head:

`group, company`
- [PLUS] Fits well with the proposed pattern (because it's the terms the author's had in mind initially)
- [PLUS] Less re-writing burden
- [MINUS] Excludes specific organizations that are not groups/companies but still consist of multiple legal entities and want to do InnerSource (private / publich research institutions, government agencies, ...)
- [MINUS] Ambiguity of both terms
- [MINUS] Less clear on describing the actual type of building block of the organization that is relevant for this pattern (which, as I read it, applies primarily to multiple legal entities)

So, my suggestion would still be, to use different terms, probably `organization` and `legal entity`

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: >  [...] AbstractSuperContext [...]

Agreed, @spier! That's a topic for elsewhere :)

> I think the specific condition for an Inner Source License is that the organization has a complexity so that there are legal contracts necessary within the organization. That probably corresponds with size and numbers of developers in most cases.

Cool. It doesn't seem to be merely the size then. A possible fix could mabye be something like that? :)

```suggestion
- A sufficient complexity of the organization and its organizational structure
```

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: > I would say that external regulations do have an impact, but not a very specific one. 

I assume you worked a lot with our legal team in designing your org's InnerSource license. Would you be open to asking them if they see specific external regulations that require that? (I understand if you don't want to: questions to the legal team can leverage unintended consequences).

If in your observation the internal regulations were the key force. Please feel free to resolve this comment @cornelius and @spier.

File: README.md
Comment (spier) [inline README.md]: Reminder to self:

Once the pattern is fully reviewed, check if we need to adapt this Patlet here as well.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: (Intended to only mark `Company culture`)

To clarify the suggestion: Make explicit how `company culture` plays into this pattern:
```suggestion
- **Company culture** that isn’t used to InnerSource collaboration and sharing code. This results in uncertainty about the rights and obligations when using shared code.
```

One could very well argue: "Whoops. Doesn't that drift too much into the problem section?" but I think it is important to have a tie in between the forces and other sections where possible.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: My guts feeling would be to still synthesize it and already add it to the resulting context. I'd like to hear @lenucksi's / other's opinion on that, though.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md:11]: My suggestion is for you, as authors, to consider whether the psychological uncertainty is something that you would want to add.

File: innersource-license.md
Comment (maxcapraro) [reply innersource-license.md]: > Accounting, profit sharing, taxes are all things which could be covered in an Inner Source license, but they also could be left to other legal constructions.
>
> Ownership of code, relicensing, liability, warranty are topics you probably would typically cover in a license.

I agree fully. This is how I have seen InnerSource licenses been used in organizations. For some organizations the tax / transfer pricing discussion is the first that makes explicit the need for a clear legal framework. Hence my question :) 

> - There are legal contracts in place which cover sharing source code but they are not standardized (so they create additional effort in negotiating and understanding for every project) or they don't allow sharing source code in an open enough sense as an Inner Source approach would suggest.
> 
> - There are no contracts in place but source code is shared informally. That might create uncertainty in cases where clarity about ownership and rights and obligations is needed.

I like this wording very much. I suggest to merge it into that section! :)

File: innersource-license.md
Comment (cornelius) [reply innersource-license.md]: > So, my suggestion would still be, to use different terms, probably organization and legal entity

I like the term `legal entity` because that's closely tied to the need of having legal agreements in place. `organization` as a general term would also work for me. No matter what terms we use, we should make sure that we have a definition somewhere which explains them in the context of this pattern.

Comment (maxcapraro) [reply innersource-license.md]: > No matter what terms we use, we should make sure that we have a definition somewhere which explains them in the context of this pattern.

+1000! Let's add those definitions to the context section

Comment (lenucksi) [reply innersource-license.md]: I'll go with @cornelius point of "what ever we use we should make sure that we have a definition somewhere". Let's use a glossary for that one. We argued that this would be useful across patterns in any case. Starting it can be as simple as creating a simple text file with "term: explanation" pairs in it and link to that, it's hypertext after all. 
We can always fancy that up later.

Other than that: If we talk about legal agreements between legal entities it might not harm calling them that. We could give an example "A legal entity in this context can be x,y,z". But effectively the point is definition of terms before usage and usage according to definition.

File: innersource-license.md
Comment (lenucksi) [reply innersource-license.md]: @cornelius or @maxcapraro Could you please turn said wording into a commit suggestion  or commit and have it merged? That way we have conserved it and ensure proper attribution.

File: innersource-license.md
Comment (lenucksi) [reply innersource-license.md:11]: Judging from my previous work and discussions with engineers and other non-legal professionals I can very much confirm that "psychological unsafety" is a thing. 
However, I can only join @spier's question here:  "Are you suggesting a change"?
If so: I'd be happy about (a) commit suggestion(s) at the right place(s).

File: innersource-license.md
Comment (lenucksi) [reply innersource-license.md]: What @spier states here is actually what I would more or less add to the resulting context.
- We could very thinly paraphrase the Known Instances and point there for more
-  and also state that research is still out on specific,reliable resultung context results.

File: innersource-license.md
Comment (lenucksi) [inline innersource-license.md]: From reading this I feel that we might be able to streamline the terms here a bit. 
I read: 
- company == subsidiary (BSH, RBPT, ...)
- large group == the entire thing (Bosch) - this is what I would know as "Konzern" in German.

I think subsidiary is actually a good, pretty unambiguously word. I'd vote to use this instead of "company" after introducing the word subsidiary.

File: innersource-license.md
Comment (lenucksi) [inline innersource-license.md]: Here, the term 'Enterprise' just emerged from the scrubs. 
Could we align it with the other terms the patterns has or explain the difference in semantics to the others?

**Review by lenucksi — COMMENTED** [review]
I think this is good stuff on an important topic. I've added of bunch of comments (unfortunately outside of the review function) and inside the review process.
Quite a few can likely be resolved by means of commit suggestions from existing discussions or just agreements, others possibly by Ctrl-F + Replace. 

Happy to merge once all discussions are resolved.

File: innersource-license.md
Comment (spier) [reply innersource-license.md:11]: For now, let's not add anything here.

Once other people/companies that are following this pattern start contributing to this, they can add **psychological uncertainty** here if that is something that they have seen in the context of this pattern.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I will merge your suggestion in. This also points to the "psychological uncertainty" of the individual developer a little bit. So at least we have a note of it in here that we can elaborate on later if we want to.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Given that 10 days have passed, I will resolve this. Can be expanded on later by other companies/contributors.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I have beefed up the resulting context just a little bit, following @lenucksi suggestion. Resolving this thread.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Have replaced it with "Large groups of companies (enterprises)" for now.

It might be that we revisit these terms completely but that is discussed at length in other comments already. Hence closing this.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Also see inline comment from @lenucksi:
https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/147/files#r417457368

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I have added this feedback to the main thread about the terminology. 

Will resolve this individual comment now.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: Based on all the feedback I will rewrite the pattern to use **organization** (for group/enterprise) and **legal entity** (for company/subsidiary). I will put those as definitions to the Context section.

Will resolve this thread here once done.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I have added the two bullets from @cornelius with slight modifications to the **Forces** section, as these are "constraints that can be changed at a cost".

Resolving this thread.

File: innersource-license.md
Comment (spier) [reply innersource-license.md]: I have rewritten the pattern accordingly.
 
I have placed the glossary at the end of this file, as I cannot claim that this is how these terms are used in other patterns. If we really want to produce a global glossary, I suggest that we do that as a separate activity outside of this PR>

@maxcapraro would you mind reviewing this, as you brought this topic up?

My observations after doing the rewrite:

- the pattern is certainly more consistent now 👍 
- it sounds much more like a legal document, and less like it is written for humans (which I personally don't like) 👎 
- there are a couple of places where replacing the word "company" without "legal entity" would make the sentences longer,  and also a bit awkward. I have left those as they were. Examples:
  - cross-company accounting
  - Company culture
  - inter-company relationships

Still I think that this was a good iteration and if this gets sign-off, I would gladly get this merged to master as it already provides value to the InnerSource community as is.

File: README.md
Comment (spier) [reply README.md]: Done

**Comment by spier** [comment]
Just to confirm the latest status down here as well:

I worked in all feedback, and resolved the open conversations.

Waiting for confirmation on the remaining open conversation above. Once that clears, this PR should be ready to merge.

**Comment by cornelius** [comment]
Thanks, @spier. The PR looks good to me now. Also thanks to everybody else who chimed in with feedback. I like the result. This is a great first iteration of the pattern.

File: innersource-license.md
Comment (gruetter) [inline innersource-license.md:32]: This one' a bit fuzzy to me. 

```suggestion
- Unregulated use of the software can lead to unclear ownership structures, duplication of improvement effort or unwanted competition
```

File: innersource-license.md
Comment (gruetter) [inline innersource-license.md:34]: This one is at least partially a duplication of the force above, regarding the consequences of _Freedom over using the software"_

File: innersource-license.md
Comment (gruetter) [inline innersource-license.md:43]: We should mention that an InnerSource license needs to be ratified by the responsible bodies in the respective organisation. If there is no such ratification, the license might not be widely accepted by management ... or developers with a high need for job security.

Comment (gruetter) [reply innersource-license.md:43]: It might also make sense to mention to not be too strict about what can be done with the SW. We (at Bosch) made that mistake and precluded the use of InnerSource software in Open Source projects later on. This turned out to be a killer criterium for some of our projects and eventually triggered the development of a new version of the license, which took a lot of effort.

File: innersource-license.md
Comment (gruetter) [inline innersource-license.md:68]: Add Bosch as an known instance, too. Here's my proposal for content:

"The InnerSource initiative at Robert Bosch GmbH was also governed with a license from the beginning in 2009. With over 500 legal entities, sharing software internally was very difficult and time consuming to implement in a compliant way. InnerSource and specifically the InnerSource license changed that completely. Today, collaboration in software projects governed by their InnerSource license is the de facto standard and widely practiced. Bosch has evolved their license over the years, e. g. to also allow the eventual publication of InnerSource software as Open Source software, which wasn't possible with early versions of their license."

**Review by gruetter — COMMENTED** [review]
Nice pattern, well done guys!  I wonder why this wasn't created earlier ;).

**Comment by spier** [comment]
@gruetter thanks for your feedback. The additional info on the "known instances" of this pattern in Bosch will certainly be interesting for the readers.

As this PR has already taken 23 days up to this point, I would like to get the current version merged, and move further modification proposals to separate PRs. e.g. As soon as this PR here is merged, you can open a new PR that adds the "known instance" portion for Bosch.

Hope this approach is ok?

@maxcapraro we got sign-off from Cornelius on the latest version of this pattern (see above). If you could review and resolve the last open conversation thread at the very top, then @lenucksi can merge this PR :)

Thanks.

**Review by maxcapraro — APPROVED** [review]
Thanks @spier and @cornelius :) All my comments were adressed. Thus, I am changing my review status on this PR to approved.

(PS: Some comments by other reviewers are still open and worth adressing)

**Comment by lenucksi** [comment]
Hi @gruetter and @maxcapraro , thanks your reviews!
I see that there are [great proposals](https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/147#discussion_r419327232) (and more) for additional content @gruetter - could you please turn them into a separate pull request to extend this pattern? And link to the new one from here?

**Pull request merged**
